c_src/mongoose_jid.c

```diff
@@ -0,0 +1,73 @@
+#include "erl_nif.h"
+#include <stdio.h>
+#include <string.h>
+
+ERL_NIF_TERM
+mk_error(ErlNifEnv* env)
+{
+    ERL_NIF_TERM ret;
+    enif_make_existing_atom(env, "error", &ret, ERL_NIF_LATIN1);
+    return ret;
+}
+
+static ERL_NIF_TERM
+from_binary_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    if(argc != 1) {
+        return enif_make_badarg(env);
+    }
+
+    const ERL_NIF_TERM arg = argv[0];
+    ErlNifBinary bin;
+    if (!enif_inspect_binary(env, arg, &bin)) {
+        return enif_make_badarg(env);
+    }
+
+    const size_t size = bin.size;
+    unsigned commercial_at = -1;
+    unsigned slash = size;
+    for (unsigned i = 0; i < size ; ++i) {
+        switch(bin.data[i]) {
+            case '/':
+                if (slash == size) {
+                    slash = i;
+                    goto end_loop;
+                    // If we found a slash first, we don't care about commercial_ats anymore
```

Comment:
We could've found a `@` before, so it's not necessarily **first**.
```suggestion
                    // If we found a slash, we don't care about commercial_ats anymore
```

Reply:
We could have, a jid with a resource but no userpart is a valid jid, even when the resource also contains `@`, see in https://datatracker.ietf.org/doc/html/rfc6122#section-2.1
```
      jid           = [ localpart "@" ] domainpart [ "/" resourcepart ]
```

And also https://datatracker.ietf.org/doc/html/rfc7622#section-3.5
```
   | 15| <a.example.com/b@example.net>| A domainpart followed by a    |
   |   |                              | resourcepart that contains an |
   |   |                              | "at" sign   
```